### PR TITLE
feat(subscription): add configurable heartbeat for subscription callback protocol

### LINKED ISSUE #4115: callback: configurable heartbeat rate for callback protocol
It would also be a way to disable heartbeat for architecture based on lambdas

### PR DESCRIPTION
**Pull request by bnjjj**
Add ability to disable heartbeat for architecture based on lambdas for example

Example of configuration:

```yaml
subscription:
  enabled: true
  mode:
    preview_callback:
      public_url: http://127.0.0.1:4000
      heartbeat_interval: 5s # Optional
      listen: 127.0.0.1:4000
      path: /callback
      subgraphs:
      - accounts
 ```

Fixes #4115

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

**Comment by bnjjj** [comment]
@dariuszkuc and @trevor-scheer I added you on this PR because you both worked on different implementation of that protocol in different languages. Let me know if it looks good to you :)

**Comment by dariuszkuc** [comment]
Should we then allow the heartbeat interval to be configurable on the subgraphs as well? (currently I got it hard coded to 5s for Spring GraphQL)

**Comment by bnjjj** [comment]
@dariuszkuc Yes, that's why I'm sending the heartbeat value in the callback extension when sending the subscription operation to the subgraph.

File: docs/source/executing-operations/subscription-callback-protocol.mdx [hunk 1]
```diff
@@ -52,6 +53,7 @@ All actions described in these steps are performed by one of **Router**, **Subgr
     - `callback_url`: The URL that **Emitter** will send subscription data to
     - `subscription_id`: The generated unique ID for the subscription operation
     - `verifier`: A string that **Emitter** will include in all HTTP callback requests to verify its identity
+    - `heartbeat_interval_ms`: The number of milliseconds a heartbeat has to be sent to the callback endpoint for the subscription operation
```
Comment (dariuszkuc) [inline docs/source/executing-operations/subscription-callback-protocol.mdx:56]: I think you should call out that this is optional? i.e. if not specified then we don't send heartbeat?

File: docs/source/executing-operations/subscription-callback-protocol.mdx [hunk 2]
```diff
@@ -273,54 +275,4 @@ The following are common error states that can occur with this protcol:
 - **Emitter** receives an error HTTP status code from **Router**'s callback endpoint. 
   - If the error code is 404, **Emitter** should consider the associated subscription terminated by **Router**.
   - In other error cases, **Emitter** should consider the subscription terminated due to an unexpected error.
-- **Emitter** fails to send **Router** a [`check`](#check) message for an active subscription every five seconds, causing **Router** to terminate that subscription.
-
-## Examples
```
Comment (dariuszkuc) [inline docs/source/executing-operations/subscription-callback-protocol.mdx:278]: i think it was nice to have a step by step example of the protocol execution

File: docs/source/executing-operations/subscription-callback-protocol.mdx (hunk 1 above)
Comment (bnjjj) [reply docs/source/executing-operations/subscription-callback-protocol.mdx:56]: Oh no you're right I didn't document the behavior when heartbeat is disabled. The logic I wrote is when the heartbeat is set to 0 it means it's disabled

File: docs/source/executing-operations/subscription-callback-protocol.mdx (hunk 2 above)
Comment (bnjjj) [reply docs/source/executing-operations/subscription-callback-protocol.mdx:278]: Actually there is already an example in the same document. Could you read it and let me know if you think it's enough ?

Comment (dariuszkuc) [reply docs/source/executing-operations/subscription-callback-protocol.mdx:278]: 👍 didn't notice the other section, yeah this is redundant

File: docs/source/executing-operations/subscription-callback-protocol.mdx [hunk 3]
```diff
@@ -90,7 +92,7 @@ With initialization complete, the protocol commences its [main loop](#main-loop)
 
 The protocol's main loop remains active for the duration of the subscription. During the main loop, all of the following occur:
 
-- Every five seconds, **Emitter** must send a [`check` message](#check) to **Router** to confirm that **Router** is still listening.
+- Every `heartbeat_interval_ms` milliseconds (value coming from the initial payload sent in `extensions` from the **Router**), **Emitter** must send a [`check` message](#check) to **Router** to confirm that **Router** is still listening.
```
Comment (dariuszkuc) [inline docs/source/executing-operations/subscription-callback-protocol.mdx:95]: maybe call out that we only do it IF heartbeat is enabled?

File: docs/source/executing-operations/subscription-callback-protocol.mdx [hunk 4]
```diff
@@ -195,7 +197,7 @@ Fields and behaviors specific to individual message types are documented below.
 
 During protocol [initialization](#initialization), **Subgraph** sends a synchronous `check` message to **Router** to help ensure that it can send callbacks successfully, and that the `id` and `verifier` fields provided by **Router** are correct.
 
-As long as subscription is active, **Emitter** must send a `check` message to **Router** every five seconds. This enables **Emitter** to confirm both that it can still reach **Router**'s callback endpoint, and that subscription is still active.
+As long as subscription is active, **Emitter** must send a `check` message to **Router** every `heartbeat_interval_ms` milliseconds (value coming from the initial payload sent in `extensions` from the **Router**). This enables **Emitter** to confirm both that it can still reach **Router**'s callback endpoint, and that subscription is still active.
```
Comment (dariuszkuc) [inline docs/source/executing-operations/subscription-callback-protocol.mdx:200]: same comment as above -> do we need to explicitly say that only do it if heartbeats are enabled?

**Comment by dariuszkuc** [comment]
FYI https://github.com/apollographql/federation-jvm/pull/360

File: docs/source/executing-operations/subscription-callback-protocol.mdx (hunk 3 above)
Comment (bnjjj) [reply docs/source/executing-operations/subscription-callback-protocol.mdx:95]: Good call

File: .changesets/feat_bnjjj_feat_4115.md [hunk 5]
```diff
@@ -0,0 +1,20 @@
+### feat(subscription): add configurable heartbeat for subscription callback protocol ([Issue #4115](https://github.com/apollographql/router/issues/4115))
+
+Add ability to disable heartbeat for architecture based on lambdas for example
```
Comment (garypen) [inline .changesets/feat_bnjjj_feat_4115.md:3]:
```suggestion
Add the ability to disable heartbeat for certain platforms; lambdas for example
```

**Review by garypen — APPROVED** [review]
Small changelog suggestion.

**Review by UncleJooe — APPROVED** [review]

**Review by UncleJooe — CHANGES_REQUESTED** [review]
review

**Pull request merged**